### PR TITLE
[Contracts][DependencyInjection] Add `ServiceCollectionInterface`

### LINKED ISSUE #19368: [Contracts][DependencyInjection] Add `ServiceCollectionInterface`
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#53163
| PR author(s) | @kbond
| Merged in    | 7.1

### PR DESCRIPTION
**Pull request by alexandre-daubois**
Fix #19368

Ping @kbond 🙂

### CODE REVIEW

File: service_container/service_subscribers_locators.rst [hunk 1]
```diff
@@ -122,6 +122,24 @@ It is also a callable and a countable::
 
     return $handler->handle($command);
 
+Because the service locator implements the
```
Comment (kbond) [inline service_container/service_subscribers_locators.rst]: I wonder if we should move this to the above paragraph where we explain what `ServiceLocator` implements?

File: service_container/service_subscribers_locators.rst [hunk 2]
```diff
@@ -122,6 +122,24 @@ It is also a callable and a countable::
...
+:class:`Symfony\\Contracts\\Service\\ServiceCollectionInterface`, it is also
+a ``\Traversable``. This allows to iterate through all registered services to
+apply some operations on them::
+
+    foreach ($this->locator as $serviceId => $service) {
+        // access to the service directly
+        $service->someMethod();
+
+        // or via the `get()` method
+        $this->locator->get($serviceId)->someMethod();
```
Comment (kbond) [inline service_container/service_subscribers_locators.rst]: Let's move this line out of the foreach as a separate example. `$service` is initialized at this point.

File: service_container/service_subscribers_locators.rst [hunk 3]
```diff
@@ -122,6 +122,24 @@ It is also a callable and a countable::
 
     return $handler->handle($command);
 
+Because the service locator implements the
+:class:`Symfony\\Contracts\\Service\\ServiceCollectionInterface`, it is also
+a ``\Traversable``. This allows to iterate through all registered services to
+apply some operations on them::
+
+    foreach ($this->locator as $serviceId => $service) {
```
Comment (kbond) [inline service_container/service_subscribers_locators.rst:135]: I also think we should merge this code block with the one above.

**Review by kbond — CHANGES_REQUESTED** [review]
Thanks for working on this Alex!

**Comment by alexandre-daubois** [comment]
Thanks for your suggestions Kevin! I reworked the PR, is it better now? 🙂

File: service_container/service_subscribers_locators.rst [hunk 4]
```diff
@@ -111,17 +111,30 @@ in the service subscriber::
...
     $numberOfHandlers = count($this->locator);
     $nameOfHandlers = array_keys($this->locator->getProvidedServices());
+
+    // you can iterate through all services of the locator
+    foreach ($this->locator as $serviceId => $service) {
+        // do something with the service, the service id or both
+    }
+
     // ...
     $handler = ($this->locator)($commandClass);
```
Comment (kbond) [inline service_container/service_subscribers_locators.rst]: `ServiceLocator` is indeed callable but this isn't part of the contracts:

```php
public function someAction(ServiceCollectionInterface $locator)
{
    $handler = ($locator)($commandClass); // this works but static analysis will complain
}
```

I'd prefer removing this in favor of `$this->locator->get($commandClass)` but as we've shown this callable example for some time, I'm not sure... @nicolas-grekas, what do you think. I don't have the context on why `ServiceLocator` was made callable initially.

Comment (nicolas-grekas) [reply service_container/service_subscribers_locators.rst]: I would reformulate the whole paragraph. Why tell that the injected service is some implementation-specific class while only an interface is type hinted? This is a bad hint that encourages against SOLID principles.

Comment (alexandre-daubois) [reply service_container/service_subscribers_locators.rst]: Alright, I tried something totally different not mentioning "internal" implementation but only explaining the locators are countable and traversable.

File: service_container/service_subscribers_locators.rst [hunk 5]
```diff
@@ -110,18 +110,30 @@ in the service subscriber::
     that you have :ref:`autoconfigure <services-autoconfigure>` enabled. You
     can also manually add the ``container.service_subscriber`` tag.
 
-The injected service is an instance of :class:`Symfony\\Component\\DependencyInjection\\ServiceLocator`
-which implements both the PSR-11 ``ContainerInterface`` and :class:`Symfony\\Contracts\\Service\\ServiceProviderInterface`.
-It is also a callable and a countable::
+The injected service is an instance of
+:class:`Symfony\\Component\\DependencyInjection\\ServiceLocator`. A service
```
Comment (nicolas-grekas) [inline service_container/service_subscribers_locators.rst]: I still wouldn't mention this, it's an implementation detail.
We should tell about ServiceCollectionInterface: if ppl type-hint for it, *then* they will get an object that is coutable+iterable, in addition to being a PSR11 locator.

**Comment by alexandre-daubois** [comment]
I tried to remove useless implementation details and go for the type-hint approach Nicolas suggested.

**Comment by javiereguiluz** [comment]
A nice contribution! Thanks a lot Alex and thanks for your patience during the review 🙏 

Thanks also to Kevin and Nicolas for the review!

**Pull request merged**